endpoints/openrtb2/auction.go

```diff
@@ -1531,8 +1531,12 @@ func (deps *endpointDeps) validateImpExt(imp *openrtb_ext.ImpWrapper, aliases ma
 		return []error{fmt.Errorf("request validation failed. The StoredAuctionResponse.ID field must be completely present with, or completely absent from, all impressions in request. No StoredAuctionResponse data found for request.imp[%d].ext.prebid \n", impIndex)}
 	}
 
-	if len(storedBidResp) > 0 {
-		if err := validateStoredBidRespAndImpExtBidders(prebid.Bidder, storedBidResp, imp.ID); err != nil {
+	if len(prebid.StoredBidResponse) > 0 || len(storedBidResp) > 0 {
+		// prebid.StoredBidResponse - data from incoming imp.ext.prebid.storedbidresponse
+		// storedBidResp - imp id mapped to bidders from imp.ext (case insensitive) that have matched stored bid response
+		// in case only one of prebid.StoredBidResponse or storedBidResp exist means stored responses are specified
+		// in imp.ext.prebid.storedresponse, but don't match bidders in imp.ext (or imp.ext.prebid.bidders)
+		if err := deps.validateStoredBidRespAndImpExtBidders(prebid.Bidder, storedBidResp, imp.ID); err != nil {
```

Comment:
This works but I think the fact that we need this comment is an indication that perhaps things could be simplified.
How about a few slight modifications like this where we always call the validate function which then does the entire analysis including the edge case we are trying to address:
```
hasPrebidStoredBidResponse := len(prebid.StoredBidResponse) > 0
if err := deps.validateStoredBidRespAndImpExtBidders(hasPrebidStoredBidResponse, prebid.Bidder, storedBidResp, imp.ID); err != nil {
	return []error{err}
}
```

```
func (deps *endpointDeps) validateStoredBidRespAndImpExtBidders(hasPrebidStoredBidResponse bool, bidderExts map[string]json.RawMessage, storedBidResp stored_responses.ImpBidderStoredResp, impId string) error {
	if storedBidResp == nil && !hasPrebidStoredBidResponse{
		return nil
	}
	if storedBidResp == nil {
		return generateStoredBidResponseValidationError(impId)
	}
	if bidResponses, ok := storedBidResp[impId]; ok {
		if len(bidResponses) != len(bidderExts) {
			return generateStoredBidResponseValidationError(impId)
		}

		for bidderName := range bidResponses {
			if _, bidderNameOk := deps.normalizeBidderName(bidderName); !bidderNameOk {
				return fmt.Errorf(`unrecognized bidder "%v"`, bidderName)
			}
			if _, present := bidderExts[bidderName]; !present {
				return generateStoredBidResponseValidationError(impId)
			}
		}
	}
	return nil
}
```

Thoughts?

Reply:
Yes, refactored. I put this part `len(prebid.StoredBidResponse) > 0` iside of the `validateStoredBidRespAndImpExtBidders` function